File: java/src/main/java/vsdk/toolkit/raycasting/common/RayTools.java
/**
Some utility routines for ray intersections and for statistics
*/

package vsdk.toolkit.raycasting.common;

import vsdk.toolkit.common.linealAlgebra.CoordinateSystem;
import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Ray;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.common.statistics.Statistics;
import vsdk.toolkit.material.PhongBidirectionalScatteringDistributionFunction;
import vsdk.toolkit.material.RayHitFlag;
import vsdk.toolkit.scene.Camera;
import vsdk.toolkit.scene.VoxelGrid;
import vsdk.toolkit.skin.Patch;
import vsdk.toolkit.skin.RayHit;

public class RayTools {
    private static int pathFrontHitFlags() {
        return RayHitFlag.FRONT | RayHitFlag.POINT | RayHitFlag.MATERIAL;
    }

    private static RayHit
    traceWorld(
        VoxelGrid sceneWorldVoxelGrid,
        Ray ray,
        Patch patch,
        int flags,
        Patch extraPatch,
        RayHit hitStore)
    {
        RayHit myHitStore = new RayHit();
        float[] dist = new float[1];
        RayHit result;
        if (hitStore == null) {
            hitStore = myHitStore;
        }

        dist[0] = Numeric.HUGE_FLOAT_VALUE;
        Patch.dontIntersect3(patch, patch != null ? patch.twin : null, extraPatch);
        result = sceneWorldVoxelGrid.gridIntersect(ray, 0.0f, dist, flags, hitStore);

        if (result != null) {
            // Compute shading frame (Z-axis = shading normal) at intersection point
            CoordinateSystem frame = result.getShadingFrame();
            result.setShadingFrame(frame);
        }
        Patch.dontIntersect0();

        return result;
    }

    public static RayHit
    findRayIntersection(
        VoxelGrid sceneWorldVoxelGrid,
        Ray ray,
        Patch patch,
        PhongBidirectionalScatteringDistributionFunction currentBsdf,
        RayHit hitStore)
    {
        int hitFlags;
        RayHit newHit;

        if (currentBsdf == null) {
            // Outside everything in vacuum
            hitFlags = pathFrontHitFlags();
        }
        else {
            hitFlags = pathFrontHitFlags() | RayHitFlag.BACK;
        }

        // Trace the ray
        newHit = traceWorld(sceneWorldVoxelGrid, ray, patch, hitFlags, null, hitStore);
        Statistics.instance().rayTracer.rayCount++; // statistics

        // Robustness test : If a back is hit, check the current
        // bsdf and the bsdf of the material hit. If they
        // don't match, exclude this patch and trace again :-(
        if (newHit != null && (newHit.getFlags() & RayHitFlag.BACK) != 0 &&
            newHit.getPatch().material.getBsdf() != currentBsdf) {
            // Whoops, intersected with wrong patch (accuracy problem)
            newHit = traceWorld(sceneWorldVoxelGrid, ray, patch, hitFlags, newHit.getPatch(), hitStore);
            Statistics.instance().rayTracer.rayCount++; // Statistics
        }

        return newHit;
    }

    /**
pathNodesVisible : send a shadow ray
*/
    public static boolean
    pathNodesVisible(
        VoxelGrid sceneWorldVoxelGrid,
        SimpleRaytracingPathNode node1,
        SimpleRaytracingPathNode node2)
    {
        Vector3D dir = new Vector3D();
        Ray ray = new Ray();
        RayHit hit;
        RayHit hitStore = new RayHit();
        double cosRay1;
        double cosRay2;
        double dist;
        double dist2;
        float[] fDistance = new float[1];
        boolean visible;
        boolean doTest;

        // Determines visibility between two nodes,
        // Returns visibility and direction from eye to light node (newDir_e)
        if (node1.m_hit.getPatch() == node2.m_hit.getPatch()) {
            // Same patch cannot see itself. Wrong for concave primitives!
            return false;
        }

        dir.subtraction(node2.m_hit.getPoint(), node1.m_hit.getPoint());

        dist2 = dir.norm2();
        dist = Math.sqrt(dist2);

        dir.inverseScaledCopy((float)dist, dir, Numeric.EPSILON_FLOAT);

        dist = dist * (1 - Numeric.EPSILON);

        ray.position = node1.m_hit.getPoint();
        ray.direction.copy(dir);

        cosRay1 = dir.dotProduct(node1.m_normal);
        cosRay2 = -dir.dotProduct(node2.m_normal);

        doTest = false;

        if (cosRay1 > 0) {
            if (cosRay2 > 0) {
                // Normal case : reflected rays.
                doTest = true;
            }
            else {
                // Node1 reflects, node2 transmits
                if (node1.m_inBsdf == node2.m_outBsdf) {
                    doTest = true;
                }
            }
        }
        else {
            if (cosRay2 > 0) {
                if (node1.m_outBsdf == node2.m_inBsdf) {
                    doTest = true;
                }
            }
            else {
                if (node1.m_outBsdf == node2.m_outBsdf) {
                    doTest = true;
                }
            }
        }

        if (doTest) {
            if (node2.m_hit.getPatch().hasZeroVertices()) {
                fDistance[0] = Numeric.HUGE_FLOAT_VALUE;
            }
            else {
                fDistance[0] = (float)dist;
            }

            Patch.dontIntersect3(
                node2.m_hit.getPatch(),
                node1.m_hit.getPatch(),
                node1.m_hit.getPatch() != null ? node1.m_hit.getPatch().twin : null);
            hit = sceneWorldVoxelGrid.gridIntersect(
                ray,
                0.0f,
                fDistance,
                RayHitFlag.FRONT | RayHitFlag.BACK | RayHitFlag.ANY,
                hitStore);
            Patch.dontIntersect0();
            visible = (hit == null);

            Statistics.instance().rayTracer.rayCount++; // Statistics
        }
        else {
            visible = false;
        }

        return visible;
    }

    /**
Can the eye see the node ?  If so, pix_x and pix_y are filled in
*/
    public static boolean
    eyeNodeVisible(
        Camera camera,
        VoxelGrid sceneWorldVoxelGrid,
        SimpleRaytracingPathNode eyeNode,
        SimpleRaytracingPathNode node,
        float[] pixX,
        float[] pixY)
    {
        Vector3D dir = new Vector3D();
        Ray ray = new Ray();
        RayHit hit;
        RayHit hitStore = new RayHit();
        double cosRayLight;
        double cosRayEye;
        double dist;
        double dist2;
        float[] fDistance = new float[1];
        double x;
        double y;
        double z;
        double xz;
        double yz;

        // Determines visibility between two nodes,
        // Returns visibility and direction from eye to light node (newDir_e)
        dir.subtraction(node.m_hit.getPoint(), eyeNode.m_hit.getPoint());

        dist2 = dir.norm2();
        dist = Math.sqrt(dist2);

        dir.inverseScaledCopy((float)dist, dir, Numeric.EPSILON_FLOAT);

        // Determine which pixel is visible
        z = dir.dotProduct(camera.Z);

        boolean visible = false;
        if (z > 0.0) {
            x = dir.dotProduct(camera.X);
            xz = x / z;

            if (Math.abs(xz) < camera.pixelWidthTangent) {
                y = dir.dotProduct(camera.Y);
                yz = y / z;

                if (Math.abs(yz) < camera.pixelHeightTangent) {
                    // Point is within view pyramid

                    // Check normal directions
                    dist = dist * (1 - Numeric.EPSILON);

                    ray.position = eyeNode.m_hit.getPoint();
                    ray.direction.copy(dir);

                    cosRayEye = dir.dotProduct(eyeNode.m_normal);
                    cosRayLight = -dir.dotProduct(node.m_normal);

                    if ((cosRayLight > 0) && (cosRayEye > 0)) {
                        fDistance[0] = (float)dist;
                        Patch.dontIntersect3(
                            node.m_hit.getPatch(), eyeNode.m_hit.getPatch(),
                            eyeNode.m_hit.getPatch() != null ? eyeNode.m_hit.getPatch().twin : null);
                        hit = sceneWorldVoxelGrid.gridIntersect(ray,
                            0.0f, fDistance,
                            RayHitFlag.FRONT | RayHitFlag.ANY, hitStore);
                        Patch.dontIntersect0();
                        // HIT_BACK removed ! So you can see through back walls with N.E.E
                        visible = (hit == null);

                        // Geometry factor
                        if (visible) {
                            if (pixX != null && pixX.length > 0) {
                                pixX[0] = (float)xz;
                            }
                            if (pixY != null && pixY.length > 0) {
                                pixY[0] = (float)yz;
                            }
                        }
                    }
                }
            }
        }

        return visible;
    }
}

File: typescript/src/main/vsdk/toolkit/raycasting/common/RayTools.ts
import { CoordinateSystem } from "../../common/linealAlgebra/CoordinateSystem";
import { Numeric } from "../../common/linealAlgebra/Numeric";
import { Ray } from "../../common/linealAlgebra/Ray";
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { Statistics } from "../../common/statistics/Statistics";
import { PhongBidirectionalScatteringDistributionFunction } from "../../material/PhongBidirectionalScatteringDistributionFunction";
import { RayHitFlag } from "../../material/RayHitFlag";
import { Camera } from "../../scene/Camera";
import { VoxelGrid } from "../../scene/VoxelGrid";
import { Patch } from "../../skin/Patch";
import { RayHit } from "../../skin/RayHit";
import { SimpleRaytracingPathNode } from "./SimpleRaytracingPathNode";

export class RayTools {
  private static pathFrontHitFlags(): number {
    return RayHitFlag.FRONT | RayHitFlag.POINT | RayHitFlag.MATERIAL;
  }

  private static traceWorld(
    sceneWorldVoxelGrid: VoxelGrid,
    ray: Ray,
    patch: Patch | null,
    flags: number,
    extraPatch: Patch | null,
    hitStore: RayHit | null
  ): RayHit | null {
    const myHitStore = new RayHit();
    const dist = [Numeric.HUGE_FLOAT_VALUE];
    const localHitStore = hitStore === null ? myHitStore : hitStore;

    Patch.dontIntersect3(patch, patch !== null ? patch.twin : null, extraPatch);
    const result = sceneWorldVoxelGrid.gridIntersect(ray, 0.0, dist, flags, localHitStore);

    if (result !== null) {
      const frame = result.getShadingFrame();
      result.setShadingFrame(frame);
    }
    Patch.dontIntersect0();

    return result;
  }

  public static findRayIntersection(
    sceneWorldVoxelGrid: VoxelGrid,
    ray: Ray,
    patch: Patch | null,
    currentBsdf: PhongBidirectionalScatteringDistributionFunction | null,
    hitStore: RayHit | null
  ): RayHit | null {
    let hitFlags: number;

    if (currentBsdf === null) {
      hitFlags = RayTools.pathFrontHitFlags();
    }
    else {
      hitFlags = RayTools.pathFrontHitFlags() | RayHitFlag.BACK;
    }

    let newHit = RayTools.traceWorld(sceneWorldVoxelGrid, ray, patch, hitFlags, null, hitStore);
    Statistics.instance().rayTracer.rayCount++;

    if (newHit !== null
      && (newHit.getFlags() & RayHitFlag.BACK) !== 0
      && newHit.getPatch() !== null
      && newHit.getPatch()!.material !== null
      && newHit.getPatch()!.material!.getBsdf() !== currentBsdf) {
      newHit = RayTools.traceWorld(sceneWorldVoxelGrid, ray, patch, hitFlags, newHit.getPatch(), hitStore);
      Statistics.instance().rayTracer.rayCount++;
    }

    return newHit;
  }

  public static pathNodesVisible(
    sceneWorldVoxelGrid: VoxelGrid,
    node1: SimpleRaytracingPathNode,
    node2: SimpleRaytracingPathNode
  ): boolean {
    const patch1 = node1.m_hit.getPatch();
    const patch2 = node2.m_hit.getPatch();
    if (patch1 !== null && patch1 === patch2) {
      return false;
    }

    const dir = new Vector3D();
    const ray = new Ray();
    const hitStore = new RayHit();
    const fDistance = [0.0];

    dir.subtraction(node2.m_hit.getPoint(), node1.m_hit.getPoint());

    const dist2 = dir.norm2();
    let dist = globalThis.Math.sqrt(dist2);
    dir.inverseScaledCopy(dist, dir, Numeric.EPSILON_FLOAT);

    dist = dist * (1.0 - Numeric.EPSILON);

    ray.position = node1.m_hit.getPoint();
    ray.direction.copy(dir);

    const cosRay1 = dir.dotProduct(node1.m_normal);
    const cosRay2 = -dir.dotProduct(node2.m_normal);

    let doTest = false;

    if (cosRay1 > 0.0) {
      if (cosRay2 > 0.0) {
        doTest = true;
      }
      else if (node1.m_inBsdf === node2.m_outBsdf) {
        doTest = true;
      }
    }
    else if (cosRay2 > 0.0) {
      if (node1.m_outBsdf === node2.m_inBsdf) {
        doTest = true;
      }
    }
    else if (node1.m_outBsdf === node2.m_outBsdf) {
      doTest = true;
    }

    let visible: boolean;
    if (doTest) {
      if (patch2 !== null && patch2.hasZeroVertices()) {
        fDistance[0] = Numeric.HUGE_FLOAT_VALUE;
      }
      else {
        fDistance[0] = dist;
      }

      Patch.dontIntersect3(
        patch2,
        patch1,
        patch1 !== null ? patch1.twin : null
      );
      const hit = sceneWorldVoxelGrid.gridIntersect(
        ray,
        0.0,
        fDistance,
        RayHitFlag.FRONT | RayHitFlag.BACK | RayHitFlag.ANY,
        hitStore
      );
      Patch.dontIntersect0();
      visible = hit === null;

      Statistics.instance().rayTracer.rayCount++;
    }
    else {
      visible = false;
    }

    return visible;
  }

  public static eyeNodeVisible(
    camera: Camera,
    sceneWorldVoxelGrid: VoxelGrid,
    eyeNode: SimpleRaytracingPathNode,
    node: SimpleRaytracingPathNode,
    pixX: number[] | null,
    pixY: number[] | null
  ): boolean {
    const dir = new Vector3D();
    const ray = new Ray();
    const hitStore = new RayHit();
    const fDistance = [0.0];

    dir.subtraction(node.m_hit.getPoint(), eyeNode.m_hit.getPoint());

    const dist2 = dir.norm2();
    let dist = globalThis.Math.sqrt(dist2);

    dir.inverseScaledCopy(dist, dir, Numeric.EPSILON_FLOAT);

    const z = dir.dotProduct(camera.Z);

    let visible = false;
    if (z > 0.0) {
      const x = dir.dotProduct(camera.X);
      const xz = x / z;

      if (globalThis.Math.abs(xz) < camera.pixelWidthTangent) {
        const y = dir.dotProduct(camera.Y);
        const yz = y / z;

        if (globalThis.Math.abs(yz) < camera.pixelHeightTangent) {
          dist = dist * (1.0 - Numeric.EPSILON);

          ray.position = eyeNode.m_hit.getPoint();
          ray.direction.copy(dir);

          const cosRayEye = dir.dotProduct(eyeNode.m_normal);
          const cosRayLight = -dir.dotProduct(node.m_normal);

          if (cosRayLight > 0.0 && cosRayEye > 0.0) {
            fDistance[0] = dist;
            const nodePatch = node.m_hit.getPatch();
            const eyePatch = eyeNode.m_hit.getPatch();
            Patch.dontIntersect3(
              nodePatch,
              eyePatch,
              eyePatch !== null ? eyePatch.twin : null
            );
            const hit = sceneWorldVoxelGrid.gridIntersect(
              ray,
              0.0,
              fDistance,
              RayHitFlag.FRONT | RayHitFlag.ANY,
              hitStore
            );
            Patch.dontIntersect0();
            visible = hit === null;

            if (visible) {
              if (pixX !== null && pixX.length > 0) {
                pixX[0] = xz;
              }
              if (pixY !== null && pixY.length > 0) {
                pixY[0] = yz;
              }
            }
          }
        }
      }
    }

    return visible;
  }
}
